[고급 레이아웃 컴포넌트] 요술토끼(김의천) 미션 제출합니다

wzrabbit opened the pull request:
# 🎯 2단계. 고급 레이아웃 컴포넌트

## 공통
- [x] 의미있는 커밋 메시지 작성
- [x] PR에 관련 없는 변경 사항이 포함 유무 확인

## 컴포넌트 요구사항
- [x] 선택한 컴포넌트의 기능 구현
  - ~`defaultSize` prop에 대해 아직 컴포넌트가 동작하지 않는 상태입니다.~ 이제 작동합니다!
- [x] npm 배포
- [x] README.md에 코드 저자로서 특히 고민한 부분과 의도 설명
- [x] 테스트 코드를 작성하였나요? (선택 사항)

## 📌 구현한 내용 설명

안녕하세요, 도담! 2단계 미션 제출합니다. 결론부터 말씀드리자면 완벽하게 해내지는 못한 상황입니다. 😢 구현할 시간이 하루쯤밖에 없었기에 코딩 테스트를 치는 기분으로 미션을 구현했습니다.
하고 싶은 것은 많은데 시간은 항상 촉박하군요.

제가 구현해 본 컴포넌트는 `SplitPane` 컴포넌트입니다. 한정된 시간 동안 최대한 컴포넌트를 구현해 보기 위해 기본적인 기능에 집중했습니다. `prop` 으로 받는 값은 미션에 제시된 컴포넌트와 동일합니다.
이렇게까지 굳이 어려운 컴포넌트를 골라 구현을 도전해 본 이유는 제가 드래그를 다루는 로직을 구현해 보는 것이 이번이 처음이기 때문입니다. 그래서 `SplitPane` 컴포넌트에 관심이 갔고, 온갖 방법을 써서든 드래그를 구현해 보려고 노력하였습니다.

주로 한 작업이 기능이고, 커스텀 훅 함수에서 대부분의 로직을 다루므로, 여기서는 제가 구현했던 커스텀 훅에 대해 주로 이야기를 하고자 합니다.

크기를 조절할 수 있는 막대 형태의 컴포넌트를 뭐라 불러야 할 지 모르겠어서, 그냥 스크롤 바라고 부르겠습니다.

`SplitPane` 의 스크롤 바를 배치하고, 보다 자유로운 스크롤이 가능하도록 만들기 위해 스크롤 바의 `position` 을 `absolute` 로 설정해 두었습니다. 이후, 아래의 정보를 바탕으로 스크롤바가 있어야 하는 좌표가 어떤지를 계산했습니다.

### 부모 요소의 너비와 높이 (`useElementSize`)  
먼저 스크롤 바는 부모 밖으로 나가서는 안 되며, 사이즈가 `%` 형태로 들어올 경우 부모 요소의 너비와 높이에 대한 정보를 알아야만 한다고 판단했습니다. 그래서 해당 커스텀 훅을 구현하였습니다.

### 특정 요소가 클릭 중인지 확인 (`useIsHolding`)   
스크롤 바를 드래그하기 위한 동작을 진행하기 위해서는 먼저 스크롤 바가 클릭되어 있는 상태인지를 판단할 수 있어야 합니다. 그래서 해당 커스텀 훅을 구현하였습니다.

### 특정 요소를 기준으로 마우스 포인터의 상대좌표 확인 (`useOffsetMousePosition`)  
컴포넌트를 기준으로 마우스의 상대좌표를 얻어 스크롤 바가 위치해야 하는 부분을 알아내기 위해 해당 커스텀 훅을 구현하였습니다.

### `SplitPane` 에서 스크롤 바가 위치해야 하는 좌표 확인 (`useSplitPaneResizer`)  
위에서 설명한 세 개의 커스텀 훅의 도움을 받아 해당 기능을 구현합니다.

결론적으로, 커스텀 훅을 이용하여 어느 위치에 스크롤 바가 위치해야 하는지를 알아냈습니다. 그러다 보니 리랜더링이 상당히 많아 성능이 떨어지는 문제를 어쩔 수 없이 마주하게 되었습니다.
도담이 만약 `SplitPane` 과 비슷한 컴포넌트를 다루어 본 적이 있고, 보다 간편한 방법이나 팁을 알고 계시다면, 부디 공유해 주셨으면 좋겠습니다. 감사합니다!


리뷰 이후 고도화를 해 볼 점이 많아 보이니 되는대로 고도화를 시도해 보겠습니다. 활약을 기대해 주세요! 👍🏻 

## 🔗 참고 링크 (선택 사항)
- [npm 배포 페이지](https://www.npmjs.com/package/amethyst-layout)
- [Storybook](https://wzrabbit.github.io/layout-component/)

wzrabbit committed:
♻️ refactor: stringifySize 유틸 함수 이동 및 기존 컴포넌트의 import 경로 변경

wzrabbit committed:
✨ feat: 부모 요소의 크기를 바탕으로 자기 자신의 크기를 정수로 반환하는 유틸 함수 구현

wzrabbit committed:
✨ feat: 2차원 좌표를 다루기 위한 Point 타입 추가

wzrabbit committed:
✨ feat: 특정 요소의 width/height를 얻을 수 있는 커스텀 훅 구현

wzrabbit committed:
✨ feat: 특정 요소를 클릭 중인 상태인지를 얻을 수 있는 커스텀 훅 구현

wzrabbit committed:
✨ feat: 특정 요소를 기준으로 마우스 포인터의 상대좌표를 얻을 수 있는 커스텀 훅 구현

wzrabbit committed:
✨ feat: SplitPane의 크기 조절 막대가 랜더링되어야 하는 위치를 반환하는 커스텀 훅 구현

wzrabbit committed:
✨ feat: SplitPane 컴포넌트 구현

wzrabbit committed:
🧪 test: SplitPane에 대한 스토리북 추가

wzrabbit committed:
📦 chore: styled-components 설치

wzrabbit committed:
📦 chore: storybook 배포 워크플로우 제거

D0Dam commented on src/lib/components/SplitPane/SplitPane.tsx:
```diff
@@ -0,0 +1,60 @@
+          ref={resizerRef}
+        />
+      </S.Root>
+    </S.Container>
```
음... 한가지 의견을 달자면 splitPane에 다음과 같이 Container 내부에 Root내부에 2개의 태그가 기본적으로 들어가는 구조가 지금 구조인데요..!

ref 때문이라면 Container에 바로 ref를 주어도 될 것 같은데... 혹시 Root 컴포넌트가 존재해야만 하는 이유가 있을까요? Container에서 다 처리하기는 힘든 걸까요?

wzrabbit replied:
솔직히 시간이 임박했기에 임시방편으로 넣은 커밋이기는 했습니다. 하지만 이유는 있었습니다.

지금은 아니지만, 첫 제출 당시에는 PR에도 적었다시피 `absolute` 로 Resizer들의 위치를 정했다고 말씀드렸었습니다. 또한 마우스의 좌표를 바탕으로 Resizer들이 어느 위치에 있어야 하는지를 판단했다고도 말씀드렸습니다.

이 방식은 요소에 padding이 있을 경우 padding이 있는 부분도 요소의 일부로 인식해, `mousemove` 이벤트가 padding이 있는 곳까지 적용되었으며, 지금의 좌표 계산 방식으로는 padding에 해당되는 좌표를 더한 좌표로 계산되어 좌표 값이 엉망이 되는 문제가 있었습니다. 그래서 다소 더러운 방법이지만 `Container` 를 따로 둬서라도 이러한 문제를 막고 싶었습니다.

`Container` 라는 `div` 요소를 바깥쪽에 하나 더 두르면 사용자가 padding을 적용하더라도 `Container` 에만 적용될 것이고, 내부의 실제 `SplitPane` 레이아웃의 크기에는 영향을 주지 않을 테니까요...!

지금은 `absolute` 를 사용하지 않고 Flex 구조를 사용하며, 제공되는 prop도 `margin`이기 때문에 Resizer가 요소의 바깥으로 삐져나오는 문제는 발생하지 않을 것이라 생각합니다.

![image](https://github.com/woowacourse/layout-component/assets/87642422/f8092470-2cf0-4f49-a7ce-fa4e0f5beaa2)

wzrabbit replied:
이미지에서 검은색 사각형은 Resizer이고, 파란색 사각형은 `padding`을 의미합니다.

D0Dam commented on src/lib/components/SplitPane/SplitPane.tsx:
```diff
@@ -0,0 +1,60 @@
+  } = props;
+  const rootRef = useRef<HTMLDivElement>(null);
+  const resizerRef = useRef<HTMLDivElement>(null);
+  const { resizerX, resizerY } = useSplitPaneResizer({
```
개인적으로는.. 훅안에 훅안에 훅이 있는 구조라 로직을 살펴볼 때에는 좀 힘들었던 것 같아요..! 이 깊이를 좀 낮추어 보아도 좋을 것 같아요~!

wzrabbit replied:
이 부분에 대해서는 도담에게 여쭤보고 싶은 것이 있어 코멘트를 반영하지 않았습니다.

프리코스부터 함수를 사용할 때, 때로는 복잡한 함수를 더 깔끔하게 나누기 위해 복잡한 함수를 더 간단한 함수로 나누어 각 함수가 하나의 기능만 하도록 노력했던 기억이 납니다.

저는 커스텀 훅도 React의 state를 관리하는 로직이 있다는 것 외에 함수와 기능하는 것이 별반 차이가 없다고 생각했고, "특정 요소의 크기를 계산하는 커스텀 훅" 과 "특정 요소가 클릭중인 상태인지를 알려주는 커스텀 훅", 이 두 개의 훅은 유틸성이 강하다는 것을 발견했습니다.
그래서, SplitPane의 커스텀 훅이 워낙 복잡했기에 비록 depth가 늘어나더라도 커스텀 훅을 이렇게 분리하면 기존의 덩치가 큰 커스텀 훅을 분리할 수 있으며, 이후 분리된 로직들은 재사용에도 유용할 것이라 판단하여 이렇게 했습니다.

혹시 커스텀 훅 안에 커스텀 훅을 넣는 방법이 대체로 권장되지 않는 방법일까요? 그렇다면 좀 더 자세히 알려주시면 좋을 것 같습니다 👍🏻

D0Dam reviewed (changes requested):
안녕하세요 요술토끼! 예비군 이슈로 리뷰가 좀 늦었네요! 🥲 바로 본론으로 들어가겠습니다~!

우선 코드 세부적인 리뷰는 남기지 않았어요..! 요술토끼 말대로 더 고도화 할 부분이 많다고 했으니 그부분까지 볼 예정입니다! 그리고 리뷰를 위해 패키지를 다운받고 Readme 를 읽는 과정에서 누락된 게 있는 것 같더라구요..! 이 부분 밑에 적어 놓았으니 확인하고 추가 부탁드려요!

1.  요술토끼의  amethyst-layout을 0.5.0(최신 버전) 으로 다운 받았는데 안에 SplitPane이 없더라구요.. 그래서 제대로 확인을 해 볼 수가 없었어요! 새로 패키지에 추가해주세요!
2.  npm 및 Readme 파일에 SplitPane 관련 설명이 없는 것 같아요! 이 부분도 추가해 주시면 좋을 것 같아요.


요술토끼가 코드를 짜면서 고민이 많았던 것 같아요! 저도 좀 자세히 코드를 느껴(?) 보려고 직접 코드를 클론 받아 만져보았어요! 저라면 어떻게 짤지 궁금하다고 하시니.. 그에 관한 의견 전달드릴게요!

일단 요술토끼의 splitpane의 경우 안의 요소가 여러개가 아닌 2개일때만 정상적으로 동작하는 것 같더라구요. 대체적으로 splitPane의 경우 요소 2개를 가지고 영역을 나누는 방식이니 안의 요소가 2개라는 전제 조건하에 이야기를 해볼게요.(근데 사실 꼭 2개가 아니어도 이야기는 성립합니다! ㅎㅎ)
요술토끼의 훅은 외부에서 어떤 태그든 상관없이 2개를 넣고 분리할 수 있다는 점에서 큰 장점을 가진것 같아요. 이 내부에 ref 상태를 넘겨줄 수 있다면, 저는... 이 정도까지의 훅이나 로직이 필요 없다고 생각했어요! SplitPane 의 컴포넌트가 같는 JSX 코드 또한 더 간단하고 이쁜 구조로 가져갈 수 있을 것 같아요! 

저라면 useContext를 사용해서 다음과 같은 구조를 설계하고 들어갈 것 같아요.

```tsx
    <div {...props} ref={ref}>
      <splitPaneContext.Provider value={{ size, setSize }}> // 가로나 세로
        {children[0]}
        <div onMouseDown={마우스를 내렸을 때의 이벤트} /> // 구분자
        {children[1]}
      </splitPaneContext.Provider>
    </div>
```

이렇게 되면 마우스의 위치에 따라 Provider의 width 값을 지정해 줄 수 있고, 가장 상단의 ref나 children이 활용할 수 있게 되죠! 그렇게 되면 할 것은 children요소중 하나의 요소에 ref를 입혀서 dom style을 가져올수 있게끔 바꿔주면 될 것 같아요! 내부 children은 아래와 같은 로직으로 크기를 가져올 수 있겠죠?

```tsx
// width 기준,  상태값은 useContext로 정의됨.
useEffect(() => {
    if (!size) {
      setSize(ref.current.clientWidth);
      return;
    }

    topRef.current.style.width = `${size}px`;
  }, [size]);
```

그럼 마지막으로 SplitPane 가장 상단 요소에 적용해 놓은 ref를 통해 스타일을 조정하면 될 것 같네요! 이 때는  요술토끼가 가져온 방식도 사용 가능하고.. 그냥 창 전체 문서의 높이를 통해서 계산할 수도 있겠네요!
당연히! 요술토끼가 알수도있지만.. 관련 계산할 때 제가 자주보던 사진을 전달드릴게요~!

<img width="719" alt="image" src="https://github.com/woowacourse/layout-component/assets/51052049/9ac2b87c-2cac-455e-b186-25bfabe7d185">

딱 요소가 두개라는 가정이라면 다음과 같이 아예 props로 요소를 전달할 수도 있을 것 같네요~!

```tsx
<SplitPanePair left={<Pane />} right={<Pane />} />
```

이건 관련 리소스가 있길래 가져와봤어요!
참고 : https://codepen.io/justkash/pen/eeWpgm?editors=0110

이만.. 글을 마쳐볼게요! 이번에는 시간에 쫓기는 상황이 아니니.. 천천히 요술토끼의 생각을 잘 반영해주었으면 더욱 좋을 것 같네요~! 요술토끼의 고도화되고 깔끔해진 컴포넌트를 기대하고 있겠습니다!
뿅!

wzrabbit committed:
📦 chore: useElementSize, useIsHolding 커스텀 훅을 제외한 모든 SplitPane의 작업사항 삭제

wzrabbit committed:
✨ feat: Resizer 컴포넌트 구현

wzrabbit committed:
✨ feat: 주어진 정보들을 바탕으로 SplitPane의 요소들의 크기가 어떻게 되야 하는지를 계산하는 커스텀 훅 구현

wzrabbit committed:
✨ feat: Direction 타입 추가

# Conflicts:
#	src/lib/types/index.ts

wzrabbit committed:
✨ feat: 기본적인 기능을 지닌, 2개 이상의 요소를 담을 수 있는 SplitPane 컴포넌트 구현

- 2개 이상의 child를 담을 수 있음
- SplitPane의 방향을 정할 수 있음
- 기본적으로 요소들이 가지게 할 크기를 % 단위로 넣을 수 있음

wzrabbit committed:
✨ feat: 요소의 최소 비율과 최대 비율을 정할 수 있도록 개선

wzrabbit committed:
✨ feat: Resizer 디자인 및 테마 적용이 가능하도록 개선

wzrabbit committed:
♻️ refactor: 드래그가 부드럽게 진행되도록 개선

- 기존 드래그 이벤트에 preventDefault를 추가

wzrabbit committed:
✨ feat: border의 크기를 정할 수 있도록 개선

wzrabbit committed:
✨ feat: margin의 크기를 정할 수 있도록 개선

wzrabbit committed:
♻️ refactor: 마우스를 뗐을 때 불필요하게 모든 컴포넌트가 리랜더링 되지 않도록 개선

wzrabbit committed:
✨ feat: width, height를 정할 수 있도록 개선하고, 다양한 사이즈 단위 사용가능

wzrabbit committed:
🧪 test: SplitPane에 대한 스토리북 추가

wzrabbit committed:
♻️ refactor: 커스텀 훅의 경로 변경

wzrabbit committed:
✨ feat: SplitPane 컴포넌트가 사용 가능하도록 index 파일 수정

wzrabbit committed:
📦 chore: 불필요한 useIsHolding 커스텀 훅 제거

wzrabbit committed:
🧪 test: SplitPane 스토리북에 width/height 커스터마이징울 다루는 스토리 추가

wzrabbit committed:
📦 chore: 불필요한 asterisk 제거

wzrabbit committed:
📚 docs: SplitPane에 대한 README 문서 추가

wzrabbit committed:
✨ feat: Resizer에 hover했을 때 나오는 커서 모양을 더 어울리도록 변경

wzrabbit committed:
📦 chore: 버전 변경 - v0.5 -> v0.6

wzrabbit commented:
<img src="https://github.com/woowacourse/layout-component/assets/87642422/a88e7ff0-8c59-4046-85e3-669c52d9a55c" width="500px" />

<img src="https://github.com/woowacourse/layout-component/assets/87642422/099ca970-ef99-42a3-9cf3-8042c6ecfd43" width="500px" />


안녕하세요, 도담! 다시 돌아왔습니다.
추석 동안 도담이 제공해 주신 요소의 위치에 대한 이미지표를 보면서 컴포넌트를 이리저리 만져볼 수 있었습니다. 그리고 도담이 부모 컴포넌트의 ref만으로 처리하는 방식을 보고 이벤트 위임이 생각나 레벨 1을 돌아보며 이벤트 위임을 이 미션에 어떻게 활용할 수 있을지 고민해 보았습니다.

> 당연히! 요술토끼가 알수도있지만..


허허 저는 몰랐습니다. 알려주셔서 감사합니다.

미흡한 제출물이었음에도 조언해 주셔서 정말 감사합니다.

이틀 정도의 시간을 들여, 기존의 컴포넌트를 갈아엎고 새로운 방식으로 다시 `SplitPane` 컴포넌트를 만들 수 있었습니다. 아래에 제가 무엇을 했는지 적어보겠습니다.

## 방식 변경: `absolute` 대신, Flex ⚡
`React.Children`의 `toArray()` 를 사용하면 `children` 을 나누어 관리할 수 있다는 사실을 알고 이를 적용했습니다. 이렇게 나눈 `children` 은 FlexBox에 넣었고, 각 요소 사이사이에 Resizer를 넣었습니다. `children` props가 주어지면 그걸 그대로 가져다 쓰기만 했었는데 이걸 분리해서 써 본 것은 처음입니다.

기존의 `absolute` 를 쓰는 방식과 비교했을 때, Flex를 사용하는 방식은 아래의 장점이 있었습니다.

1. 훨씬 안정적입니다.
  - 부모 요소 바깥으로 이탈하기 쉬운 `absolute` 가 적용된 컴포넌트와 달리 Flex 컴포넌트의 자식은 속성 설정만 적절히 해 준다면 부모 요소 바깥으로 나가게 될 일이 없습니다.

2. Resizer의 두께를 바탕으로 랜더링되어야 하는 요소들의 크기를 계산할 필요가 없어집니다. Resizer의 두께는 props에서 정해주고, `flex-shrink: 0` 을 주어 Resizer가 줄어들지 않도록 만들면, 나머지 요소들의 크기는 Flex에 의해 알아서 계산됩니다. 커스텀 훅이 할 일이 줄어듭니다.

3. 부모 요소의 `width` 와 `height` 의 값이 랜더링 초기에 필요하지 않게 됩니다. 별 것 아닌 것으로 보일 수도 있습니다만 이 첫 랜더링 때 `width` 와 `height` 값을 적용해 주고 그 다음부터 `width`, `height` 값이 업데이트되는 상황에서는 무시해야 하는 로직이 있었는데 이 로직이 매우 복잡했습니다. 항상 느끼지만 **초기에만 특정 값을 지정해 주는 로직** 을 구현하는 게 React에선 참 어려운 것 같습니다. 초기에만 특정 값을 지정해 주어야 하는데 그 특정 값을 바로 가져올 수 없는 상황이라면? 상상만 해도 머리가 아파옵니다...

그래서 이를 바탕으로 Flex를 이용한 방식으로 변경하였고, 그 대신 trade-off로써 각 자식 요소의 크기는 오로지 `%` 로만 정할 수 있도록 강제해야 했습니다.
하지만, 이 컴포넌트의 크기가 변하는 것에 대응할 수 있으려면 `%` 단위로 요소의 크기를 할당하도록 사용자에게 강제하는 것은 불가피하다고 생각했습니다. 컴포넌트의 크기가 변하는 것에 대응하는 것이 매우 중요한 게, 이 컴포넌트는 중첩으로 컴포넌트를 사용할 수 있는 것을 허용하기 때문에 그 특성상 컴포넌트의 크기가 다른 컴포넌트를 어떻게 조작하느냐에 따라 매우 빈번하게 변경됩니다.

## 기능 추가 ✨
#### 이제 사용자가 자식 컴포넌트를 **3개 이상** 사용할 수 있습니다.
<img src="https://github.com/woowacourse/layout-component/assets/87642422/7b12e5ef-c7de-4a57-9dd6-3e97c25e75b2" width="500px" />

- 이 때 초기 비율을 반드시 자식 컴포넌트의 개수만큼 정수 배열로 전달해야 합니다. 예를 들면, `[20, 30, 50]` 과 같이요.
- Resizer가 여러 개가 될 수 있기에, Resizer끼리 충돌하는 경우도 고려해야 했습니다. 아래의 요구 사항을 모두 충족할 수 있는 로직이 필요했습니다. 저는 자체적인 알고리즘을 사용하여 이를 해결했습니다. React나 JavaScript에서 제공해주는 더 좋은 해결책을 딱히 찾지 못했거든요. 커스텀 훅의 함수에 이 로직이 구현되어 있습니다만 이해가 안 되는 부분이 혹시나 있다면 따로 알려드릴 수 있도록 해 보겠습니다.
  -  Resizer가 충돌했음에도 계속해서 이동하는 경우, 해당하는 Resizer들은 **같이 밀려야**한다.
  - 더 이상 자리가 없어 다른 Resizer들이 같이 밀리는 것조차 불가능한 경우, **이동 자체를 막아야** 한다.
  - Resizer를 이동하게 되면 Resizer를 기준으로 한쪽에 있는 컴포넌트들은 길이가 늘어나고, 한쪽은 줄어든다. 위의 두 요구사항을 만족시키면서, **사용자가 정해둔 요소의 최소 크기 / 최대 크기의 범위를 벗어나는지를 판단하여 범위를 벗어나지 않도록** 해야 한다.


#### Resizer를 훨씬 부드럽게, 그리고 자연스럽게 움직일 수 있습니다.
- 커스텀 훅을 확인하시면 `mousedown` 이벤트는 `rootRef` 의 `current` 에 달려있는데, `mousemove` 와 `mouseup` 이벤트는 `document` 에 달려있는 것을 눈치채셨을 것입니다. 이는 `mousedown` 이벤트가 발생할 때는 어떤 Resizer가 눌렸는지를 정확하게 확인하기 위함이고, 이후 어떤 Resizer가 눌렸는지가 확실한 상황에서는 `document` 에 리스너를 달아 **사용자가 Resizer를 `SplitPane` 컴포넌트 바깥으로 움직이더라도 이벤트가 끊기지 않고 자연스럽게 움직이도록 개선했습니다**.
- `mousemove` 와 `mouseup` 이 `rootRef` 의 `current` 에 달려 있으면 마우스 포인터가 컴포넌트 안에 있을 때에 한하여 올바르게 작동합니다. 이는 사용자가 Resizer를 움직이다 마우스 포인터가 컴포넌트의 바깥으로 움직였을 경우 이벤트가 끊기거나, 컴포넌트 외부에서 마우스를 놓고 다시 컴포넌트 내부로 돌아올 때 마우스를 뗀 상태에서도 마우스를 계속 누르고 있는 것처럼 잘못 작동되는 문제가 있었습니다.
- 각 `Resizer` 에 모두 ref를 다는 것은 비효율적이라고 생각했습니다. 그래서 최상위 컴포넌트에만 이벤트를 달고 운용하는 방식인 이벤트 위임을 사용했습니다.

#### 컴포넌트를 중첩해서 사용할 수 있습니다.
- 이를 이용하면 컴포넌트의 활용도를 극적으로 올릴 수 있으며 사용자가 원하는 복잡한 레이아웃들을 구현할 수 있게 됩니다.

#### 테마를 이용하여 커스터마이징을 할 수 있습니다.
솔직히 지금까지 `amethyst-layout` 라는 라이브러리는 닉값을 전혀 못 하고 있었습니다. 이름만 그럴듯하지 관련되는 기능을 제공하는 것이 하나도 없었기 때문입니다. 이번 개선사항에서 닉값을 할 수 있도록 개선했습니다.

- 개발자로써 이 컴포넌트를 개발한 제 1의 의도는 **디자인**입니다. 사용자가 잘 꾸며진 `SplitPane` 컴포넌트를 별도의 노력 없이 사용할 수 있게 하는 것이 목표였습니다.
- `amethyst-layout` 은 색상을 정할 수 있는 기능을 직접적으로 제공하지 않습니다. 그 대신, `themeColor` 라는 속성을 통해 컴포넌트의 디자인을 정할 수 있습니다. 사용자의 자유를 조금 제한하는 대신 일관성 있고 잘 꾸며진 레이아웃을 내놓는 것도 좋을 것이라 생각했습니다.
- `themeColor` 가 지정되면 사용자가 정한 색에 맞게 자동으로 컴포넌트들의 색상이 할당됩니다. 지정하지 않았다면, `amethyst-layout` 의 닉값에 맞게 색상은 자수정과 비슷한 보라색으로 결정됩니다.

## 제출물 링크
[npm 배포 페이지](https://www.npmjs.com/package/amethyst-layout)
[Storybook](https://wzrabbit.github.io/layout-component/)
[README 설명](https://github.com/wzrabbit/layout-component/tree/amethyst-ii#splitpane)

저는 문서를 작성하는 것을 엄청나게 좋아하는데 이번에 새로 작성한 문서는 도담이 어떻게 생각하시는 지 모르겠군요.

## 코멘트
도담이 달아주신 코멘트는 모두 읽어봤으며 고민 후 코멘트를 달아보았습니다. 부디 도담의 의견도 알려주셨으면 좋겠습니다.

저는 여기서 물러나겠습니다. 도담, 감사합니다!

D0Dam commented on src/lib/components/SplitPane/Resizer/Resizer.styled.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+`;
+
+export const Diamond = styled.span<{ $size: number; $themeColor: HexColor }>`
+  display: inline-block;
+
+  width: ${({ $size }) => `${$size}px`};
+  height: ${({ $size }) => `${$size}px`};
+
+  background-color: ${({ $themeColor }) => $themeColor};
+  transform: rotate(45deg);
+`;
```
다이아몬드  스타일이 이거였군요 ㅎㅎ 디자인 너무 좋았습니다~~!

D0Dam commented on src/lib/components/SplitPane/Resizer/Resizer.tsx:
```diff
@@ -0,0 +1,28 @@
+      $themeColor={themeColor}
+      data-resizer-id={id}
+    >
+      {Array.from({ length: 3 }).map(() => (
+        <S.Diamond $size={thickness / 3} $themeColor={themeColor} />
+      ))}
```
요렇게 쓰는 거하고 차이가 있을까요?

```suggestion
      {Array.from({ length: 3 }, () => (<S.Diamond $size={thickness / 3} $themeColor={themeColor} />)}
```

추가적으로 꼭 필요하다고 볼수는 없지만 id 값도 전달해줬으면 좋겠네요!

D0Dam commented on src/lib/components/SplitPane/SplitPane.stories.tsx:
정성스러운 스토리북 문서에 감탄하고 갑니다.👍👍👍👍

D0Dam commented on src/lib/components/SplitPane/useSplitPane.ts:
```diff
@@ -0,0 +1,190 @@
+  mousePoint: { x: number; y: number };
+};
+
+const useSplitPane = (props: UseSplitPaneProps) => {
```
움직임 계산을 위해 드는 계산 로직은 이 정도는 어쩔 수 없나봐요 ㅎㅎ.. 정말 수고하셨습니다

D0Dam commented on src/lib/components/SplitPane/useSplitPane.ts:
```diff
@@ -0,0 +1,190 @@
+      let absMouseDiffRatio = Math.abs(mouseDiffRatio);
+      let movedMouseDiffRatio = 0;
+
+      if (moveDirection === 'right') {
+        absMouseDiffRatio = Math.min(
+          absMouseDiffRatio,
+          maximumRatio - calculatedElementRatios[resizerId]
+        );
+
+        for (let i = resizerId + 1; i < calculatedElementRatios.length; i++) {
+          const currentMovedDiffRatio = Math.min(
+            calculatedElementRatios[i] - minimumRatio,
+            absMouseDiffRatio - movedMouseDiffRatio
+          );
+
+          movedMouseDiffRatio += currentMovedDiffRatio;
+          calculatedElementRatios[i] -= currentMovedDiffRatio;
+        }
+
+        calculatedElementRatios[resizerId] += movedMouseDiffRatio;
+      } else {
```
이 부분의 로직은 중복되는 것 같은데, 아마 인자가 2~3개인 함수로 표현 해 줄수도 있을 것 같아요!
물론 다른곳에서는 쓰일 일이 없으니 이대로 두어도 괜찮을 것 같지만요!
혹시 "함수로 묶어볼까?" 라는 생각을 해 보셨을지 궁금합니다! 저도 개인적으로 이 이상 안쓰일 줄 알지만 코드 수를 줄이기 위한 함수를 만드는 것이 어떨지 요술토끼 생각을 들어보고 싶어졌어요.

D0Dam commented on src/lib/components/SplitPane/useSplitPane.ts:
```diff
@@ -0,0 +1,190 @@
+      return;
+    }
+
+    rootElement.addEventListener('mousedown', handleMouseDown);
+    document.addEventListener('mousemove', handleMouseMove);
+    document.addEventListener('mouseup', handleMouseUp);
+    rootElement.addEventListener('dragstart', handleIgnoreEvent);
+    rootElement.addEventListener('drop', handleIgnoreEvent);
```
언뜻보면 이벤트가  많아보이지만 전 정말 라이브러리답게 촘촘히 이벤트를 상세히 잘 주었다고 생각해요! 따봉입니다!

D0Dam commented on src/lib/hooks/useElementSize.ts:
```diff
@@ -0,0 +1,39 @@
+  height: number;
+};
+
+const useElementSize = (ref: RefObject<HTMLElement>) => {
```
resizeObserver를 사용한 훅이네요! 깔끔하니 좋습니다~!

D0Dam commented on src/lib/types/index.ts:
```diff
@@ -10,3 +10,7 @@ export type Gap =
     }
   | { gap?: Size };
+
+export type Direction = 'row' | 'column';
```
타입이 Direction이면 저는 왼, 오, 위, 아래가 생각이 나요. 지금도 물론 괜찮지만 행과 열을 표현하는 다른 네이밍이 없을까요..?
저도 고민해보다가 마땅히 안떠오르네요 ㅎㅎ.. 그래서 저였으면 이 타입은 그냥 `'row' | 'column'` 이렇게 썻을 것 같기도 합니다..

D0Dam reviewed (approved):
안녕하세요 요술토끼! 리뷰하러 돌아왔습니다!

요술토끼가 좀 바빠보여서 사알짝! 걱정을 했는데.. 이게 웬걸! 정말 잘하셨더라구요! 보고 정말 감탄했습니다..!! 이정도로 넓게 쓸 수 있게 만들어주실 줄은 몰랐어요! 로직도 최대한 계산과 사이즈 조정 로직을 제외하면 css 쪽으로 위임한 것도 best, ref와 resizeObserver를 사용한 것도 best 입니다! 제 개인적으로는 넘 잘했다고 말하고싶네요!

그런만큼 저도 큰 리뷰는 없습니다..ㅎㅎ 그냥 요술토끼의 의견이 궁금한 부분과 칭찬들을 적어놓아보았어요! ㅎㅎ 한번 참고해보시고 궁금한 부분은 언제든 다시 코멘트 해 주세요! 수고하셨습니다~!
